src/sqlfluff/utils/functional/context.py

```diff
@@ -0,0 +1,54 @@
+"""Define FunctionalContext class."""
+
+from sqlfluff.utils.functional import Segments
+from sqlfluff.core.rules import RuleContext
+
+
+class FunctionalContext:
```

Comment:
I have a possible alternative to consider for how we expose this:

For crawl behavior, there are multiple implementations, and a rule "declares" what it wants by setting the class-level `crawl_behaviour` field.

Should we consider doing something similar for rule context? In some sense, this would make the two variants of `RuleContext` "equal" in the sense of being equally easy for rules to use.

To be clear, the PR as written seems reasonable, but from a more philosophical point of view, it's interesting to consider. If we wanted to do this, it occurs to me that it may complicate the type annotations for the `_eval()` function (because the type of the "context" parameter is no longer a given).

Reply:
Based on the slack convo, I think the based approach for this is probably to go with a decorator, which additionally provides the functional context.

```
@functional_context
def _eval(self, context, functional):
...
```

It won't be crazy hard to add at this stage so I'll see if I can incorporate it into this PR.

Reply:
> Go with a decorator for sure, but use the decorator to subclass the context :slightly_smiling_face:

Reply:
On second thoughts I'm going to make an issue and approach that separately. This PR is also hella big.

Reply:
https://github.com/sqlfluff/sqlfluff/issues/3745